feat: add rbac claim auth hook. (#21263)
supabase/supabase: apps/docs/content/guides/auth/custom-claims-and-role-based-access-control-rbac.mdx

---
id: 'custom-claims-and-role-based-access-control-rbac'
title: 'Custom Claims & Role-based Access Control (RBAC)'
description: 'Use Auth Hooks to add custom claims for managing role-based access control.'
---

Custom Claims are special attributes attached to a user that you can use to control access to portions of your application. For example:

```json
"app_metadata": {
  "plan": "TRIAL",
  "user_level": 100,
  "group_name": "Super Guild!",
  "joined_on": "2022-05-20T14:28:18.217Z",
  "group_manager": false,
  "items": ["toothpick", "string", "ring"],
}
```

> While there are many ways to implement `custom claims` we recommend utilizing the [custom access token auth hook](/guides/auth/auth-hooks#hook-custom-access-token), which runs before a token is issued and allows you to add additional claims to the user's JWT.

Reviewer: ```suggestion
To implement Role-Based Access Control (RBAC) with `custom claims`, use a [Custom Access Token Auth Hook](/guides/auth/auth-hooks#hook-custom-access-token). This hook runs before a token is issued. You can use it to add additional claims to the user's JWT.
```

This guide uses the [Slack Clone example](https://github.com/supabase/supabase/tree/master/examples/slack-clone/nextjs-slack-clone) to demonstrate how to add a `user_role` claim and use it in your [Row Level Security (RLS) policies](/guides/auth/row-level-security).

## Create a table to track user roles and permissions

> In this example we're implementing two user roles with specific permissions:

Reviewer: ```suggestion
In this example, you will implement two user roles with specific permissions:
```

- `moderator`: A moderator can delete all messages but not channels.
- `admin`: An admin can delete all messages and channels.

```sql supabase/migrations/init.sql
-- Custom types
create type public.app_permission as enum ('channels.delete', 'messages.delete');
create type public.app_role as enum ('admin', 'moderator');

-- USER ROLES
create table public.user_roles (
  id        bigint generated by default as identity primary key,
  user_id   uuid references public.users on delete cascade not null,
  role      app_role not null,
  unique (user_id, role)
);
comment on table public.user_roles is 'Application roles for each user.';

-- ROLE PERMISSIONS
create table public.role_permissions (
  id           bigint generated by default as identity primary key,
  role         app_role not null,
  permission   app_permission not null,
  unique (role, permission)
);
comment on table public.role_permissions is 'Application permissions for each role.';
```

<Admonition type="note">
  

>   For the [full schema](https://github.com/supabase/supabase/blob/master/examples/slack-clone/nextjs-slack-clone/README.md) please refer to the example application on [Github](https://github.com/supabase/supabase/tree/master/examples/slack-clone/nextjs-slack-clone).

Reviewer: ```suggestion
  For the [full schema](https://github.com/supabase/supabase/blob/master/examples/slack-clone/nextjs-slack-clone/README.md), see the example application on [GitHub](https://github.com/supabase/supabase/tree/master/examples/slack-clone/nextjs-slack-clone).
```

</Admonition>

> You can now manage your roles and permissions in SQL, e.g. to add the mentioned roles and permissions from aboove you can say:

Reviewer: ```suggestion
You can now manage your roles and permissions in SQL. For example, to add the mentioned roles and permissions from above, run:
```

```sql supabase/seed.sql
insert into public.role_permissions (role, permission)
values
  ('admin', 'channels.delete'),
  ('admin', 'messages.delete'),
  ('moderator', 'messages.delete');
```

> ## Create auth hook to apply user role

Reviewer: ```suggestion
## Create Auth Hook to apply user role
```

> The [custom access token auth hook](/guides/auth/auth-hooks#hook-custom-access-token) runs before a token is issued allowing you to add additional claims to the user's JWT. You should add the custom claim to the user's `app_metadata` to make sure that the logged in user can't modify this value themselves!

Reviewer: ```suggestion
The [Custom Access Token Auth Hook](/guides/auth/auth-hooks#hook-custom-access-token) runs before a token is issued. You can use it edit the JWT. You should add your custom claim to the user's `app_metadata` to make sure that the logged-in user can't modify this value themselves!
```

Should it be `raw_app_metadata`? I have only vague memories of seeing an issue about this so I could be wrong

language plv8
as $$
  var user_role;

  // Fetch the current user's user_role from the public user_roles table.
  var result = plv8.execute("select role from public.user_roles where user_id = $1", [event.user_id]);
  if (result.length > 0) {
    user_role = result[0].role;
  } else {
    // Assign null
    user_role = null;
  }

  // Check if 'claims' exists in the event object; if not, initialize it
  if (!event.claims) {
    event.claims = {};
  }
  if (!event.claims.app_metadata) {
    event.claims.app_metadata = {};
  }

  // Update the level in the claims
  event.claims.app_metadata.user_role = user_role;

  return event;
$$;

grant usage on schema public to supabase_auth_admin;

grant execute
  on function public.custom_access_token_hook
  to supabase_auth_admin;

revoke execute
  on function public.custom_access_token_hook
  from authenticated, anon;

grant all
  on table public.user_roles
to supabase_auth_admin;

revoke all
  on table public.user_roles
  from authenticated, anon;

create policy "Allow auth admin to read user roles" ON public.user_roles
as permissive for select
to supabase_auth_admin
using (true)
```

</TabPanel>
</Tabs>

<Admonition type="note">
  

> To learn more about auth hooks, head over the the [Auth Hooks docs](/guides/auth/auth-hooks).

Reviewer: ```suggestion
To learn more about Auth Hooks, see the [Auth Hooks docs](/guides/auth/auth-hooks).
```

</Admonition>

## Using roles in RLS policies

> To utilize role-based access control (RBAC) in Row Level Security (RLS) policies, you can create an `authorize` methods that reads the user's role from their JWT and check's the role's permissions:

Reviewer: ```suggestion
To utilize Role-Based Access Control (RBAC) in Row Level Security (RLS) policies, create an `authorize` method that reads the user's role from their JWT and checks the role's permissions:
```

```sql supabase/migrations/init.sql
create function public.authorize(
  requested_permission app_permission,
  user_id uuid
)
returns boolean as $$
declare
  bind_permissions int;
begin
  select count(*)
  from public.role_permissions
  where role_permissions.permission = authorize.requested_permission
    and role_permissions.role = (auth.jwt() -> 'app_metadata' ->> 'user_role')::public.app_role
  into bind_permissions;

  return bind_permissions > 0;
end;

> $$ language plpgsql security definer;

Reviewer: question: do we need a `set search_path` for this?

```

<Admonition type="note">

You can read more about using functions in RLS policies in the [RLS guide](/guides/auth/row-level-security#using-functions).

</Admonition>

You can then use the `authorize` method within your RLS policies. For example, to enable the desired delete access, you would add the following policies:

```sql
create policy "Allow authorized delete access" on public.channels for delete using ( authorize('channels.delete', auth.uid()) );
create policy "Allow authorized delete access" on public.messages for delete using ( authorize('messages.delete', auth.uid()) );
```

> That's it, now you have a robust system in place to manage user roles and permissions within your database that automatically propagate to Supabase Auth.

Reviewer: ```suggestion
You now have a robust system in place to manage user roles and permissions within your database that automatically propagates to Supabase Auth.
```
